src/linear_algebra/matrix/symmetric.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import data.matrix.block
+
+/-!
+# Symmetric matrices
+
+This file contains the definition and basic results about symmetric matrices.
+
+## Main definition
+
+ * `matrix.is_symm `: a matrix `A : matrix n n α` is "symmetric" if `Aᵀ = A`.
+
+## Tags
+
+symm, symmetric, matrix
+-/
+
+variables {α β n m R : Type*}
+
+namespace matrix
+
+open_locale matrix
+
+/-- A matrix `A : matrix n n α` is "symmetric" if `Aᵀ = A`. -/
+def is_symm (A : matrix n n α) : Prop := Aᵀ = A
+
+lemma is_symm.eq {A : matrix n n α} (h : A.is_symm) : Aᵀ = A := h
+
+lemma is_symm.ext_iff {A : matrix n n α} : A.is_symm ↔ ∀ i j, Aᵀ i j = A i j :=
+by rw [is_symm, matrix.ext_iff]
+
+lemma is_symm.ext_iff' {A : matrix n n α} : A.is_symm ↔ ∀ i j, A j i = A i j :=
+is_symm.ext_iff
+
+lemma is_symm.apply {A : matrix n n α} (h : A.is_symm) (i j : n) : Aᵀ i j = A i j :=
+is_symm.ext_iff.1 h i j
+
+lemma is_symm.apply' {A : matrix n n α} (h : A.is_symm) (i j : n) : A j i = A i j :=
+is_symm.apply h i j
+
+lemma is_symm.ext {A : matrix n n α} : (∀ i j, Aᵀ i j = A i j) → A.is_symm :=
+is_symm.ext_iff.2
+
+lemma is_symm.ext' {A : matrix n n α} : (∀ i j, A j i = A i j) → A.is_symm :=
+is_symm.ext
+
```

Comment:
These lemmas all feel quite strange to me - are you basing them off existing lemmas elsewhere? I don't think `ext` is usually used in this way.

Reply:
I don't think I have ever used `is_symm.ext`, but `is_symm.apply'` are used in several places (not necessarily in this PR) if my memory is correct.
How about deleting the last two? @eric-wieser

Reply:
I think you maybe want to keep just these ones:
```suggestion
/-- A version of `matrix.ext_iff` that unfolds the `matrix.transpose`. -/
lemma is_symm.ext_iff {A : matrix n n α} : A.is_symm ↔ ∀ i j, A j i = A i j :=
matrix.ext_iff.symm

/-- A version of `matrix.ext` that unfolds the `matrix.transpose`. -/
@[ext]
lemma is_symm.ext {A : matrix n n α} : (∀ i j, A j i = A i j) → A.is_symm :=
matrix.ext
```

Reply:
As I see it, only your primed lemmas are useful; so let's remove the unprimed ones, leaving the names free for the primed ones to replace.

Reply:
Why don't we want
```
lemma is_symm.apply {A : matrix n n α} (h : A.is_symm) (i j : n) : Aᵀ i j = A i j :=
is_symm.ext_iff.1 h i j
```

Reply:
When I was writing, I was not sure which of ` A j i = A i j` `Aᵀ i j = A i j` will be more useful. Hence, I wrote both.

Reply:
I added back `apply` without the `ᵀ`.

Reply:
Yeah, I saw that. Could you explain why this version
```
lemma is_symm.apply {A : matrix n n α} (h : A.is_symm) (i j : n) : Aᵀ i j = A i j :=
is_symm.ext_iff.1 h i j
```
won't be useful?

Reply:
Which version is better also confused me when I was writing. @eric-wieser

Reply:
If you're resorting to proving things by adding `i` and `j`, then your next step is going to be to rewrite by `transpose_apply` etc; so we may as well shortcut that step.